File: 210. Course Schedule II/210.cpp
/*
topological sorting
time: O(V+E)
space: O(V+E)
*/
class Solution {
public:
    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
        vector<int> indegree(numCourses, 0);
        vector<vector<int>> adj(numCourses);
        vector<int> ordering;

        for(vector<int> prequisite: prerequisites){
            int a = prequisite[0];
            int b = prequisite[1];
            // b pointing towards a
            adj[b].push_back(a);
            indegree[a]++;
        }

        queue<int> q;
        for(int i = 0; i < numCourses; i++){
            if (indegree[i] == 0){
                q.push(i);
            }
        }

        while (!q.empty()){
            int curNode = q.front();
            q.pop();
            ordering.push_back(curNode);
            for(int neighbor: adj[curNode]){
                indegree[neighbor]--;
                if (indegree[neighbor] == 0){
                    q.push(neighbor);
                }
            }
        }

        if (ordering.size() == numCourses)
            return ordering;
        else{
            vector<int> emptyarray;
            return emptyarray;
        }
    }
};

File: 210. Course Schedule II/210.py
"""
graph + topological sorting 
n: numCourses
m: prerequisites
time: O(n+m)
space: O(n + m)
"""
from collections import deque

class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        adjlist = []
        for _ in range(numCourses):
            adjlist.append([])
        
        indegree_map = {}
        for i in range(numCourses):
            indegree_map[i] = 0
        
        ordering = []

        for prereq in prerequisites:
            a = prereq[0]
            b = prereq[1]
            adjlist[b].append(a)
            indegree_map[a] += 1

        queue = collections.deque([])

        for key, value in indegree_map.items():
            if value == 0:
                queue.append(key)

        while len(queue) != 0:
            current = queue.popleft()
            ordering.append(current)
            
            for neighbor in adjlist[current]:
                indegree_map[neighbor] -= 1

                if indegree_map[neighbor] == 0:
                    queue.append(neighbor)
        
        if len(ordering) == numCourses:
            return ordering
        else:
            return []
        
